crates/iceberg/src/expr/visitors/strict_projection.rs

```diff
@@ -0,0 +1,3231 @@
+            partition_spec,
+            cached_parts: HashMap::new(),
+        }
+    }
+
+    fn get_parts_for_field_id(&mut self, field_id: i32) -> &Vec<PartitionField> {
+        if let std::collections::hash_map::Entry::Vacant(e) = self.cached_parts.entry(field_id) {
+            let mut parts: Vec<PartitionField> = vec![];
+            for partition_spec_field in self.partition_spec.fields() {
+                if partition_spec_field.source_id == field_id {
+                    parts.push(partition_spec_field.clone())
+                }
+            }
+
+            e.insert(parts);
+        }
+
+        &self.cached_parts[&field_id]
+    }
+
+    pub(crate) fn strict_project(
+        &mut self,
+        predicate: &BoundPredicate,
+    ) -> crate::Result<Predicate> {
+        visit(self, predicate)
+    }
+
+    fn get_parts(
+        &mut self,
+        reference: &BoundReference,
+        predicate: &BoundPredicate,
+    ) -> Result<Predicate, Error> {
+        let field_id = reference.field().id;
+
+        // This could be made a bit neater if `try_reduce` ever becomes stable
+        self.get_parts_for_field_id(field_id).iter().try_fold(
+            Predicate::AlwaysFalse,
+            |res, part| {
+                // consider (ts > 2019-01-01T01:00:00) with day(ts) and hour(ts)
+                // projections: d >= 2019-01-02 and h >= 2019-01-01-02 (note the inclusive bounds).
+                // any timestamp where either projection predicate is true must match the original
+                // predicate. For example, ts = 2019-01-01T03:00:00 matches the hour projection but not
+                // the day, but does match the original predicate.
+                Ok(
+                    if let Some(pred_for_part) =
+                        part.transform.strict_project(&part.name, predicate)?
+                    {
+                        if res == Predicate::AlwaysFalse {
+                            pred_for_part
+                        } else {
+                            res.or(pred_for_part)
+                        }
+                    } else {
+                        res
+                    },
+                )
+            },
+        )
+    }
+}
+
+impl BoundPredicateVisitor for StrictProjection {
+    type T = Predicate;
+
+    fn always_true(&mut self) -> crate::Result<Self::T> {
+        Ok(Predicate::AlwaysTrue)
+    }
+
+    fn always_false(&mut self) -> crate::Result<Self::T> {
+        Ok(Predicate::AlwaysFalse)
+    }
+
+    fn and(&mut self, lhs: Self::T, rhs: Self::T) -> crate::Result<Self::T> {
+        Ok(lhs.and(rhs))
+    }
+
+    fn or(&mut self, lhs: Self::T, rhs: Self::T) -> crate::Result<Self::T> {
+        Ok(lhs.or(rhs))
+    }
+
+    fn not(&mut self, _inner: Self::T) -> crate::Result<Self::T> {
+        panic!("InclusiveProjection should not be performed against Predicates that contain a Not operator. Ensure that \"Rewrite Not\" gets applied to the originating Predicate before binding it.")
```

Comment:
How about returning errors instead of `panic`?

Reply:
Seems this is a compile-time invariant.
https://github.com/apache/iceberg-python/blob/0aecc0365b08617deee97d8a88c16d5e3aa0724e/pyiceberg/expressions/visitors.py#L810

We should call `rewrite_not` in the project function to ensure this invariant at compile time. It also needs to be done at InclusiveProjection.

Reply:
So it's better to use `unreachable!()` here?

Reply:
I think there is a discussion about this in: https://github.com/apache/iceberg-rust/pull/335/files#r1571889478 and it explain why we don't call `rewrite_not` inside project function.  But this design may cause some problems because in the future the InclusiveProjector may not only used in table scan and the new development may not realized this point. I think at least we should add comment for project to declare that the predicate must not contains `not`. And we can provide two way: 
```
fn project_predicate_without_not() // project predicate without not 
fn project() // rewrite the predicate before project 
```
cc @sdd @Fokko @marvinlanhenke

Reply:
I always thought it would be nice to be able to capture this invariant in the type system somehow but it is probably more work than it is worth and adds more complexity than we'd gain in value :-)

Personally I'd have thought the combination of a doc comment on the struct and the `unreachable!` in the not-handler should be sufficient for anyone who is doing any reasonable level of testing to be able to deal with this during their development process. We can always revisit it if it ever does cause problems in the future rather than spending time on it now.

Reply:
Yes, it should not happen since all the expressions should be rewritten by the time it reaches this visitor. We just have to make sure that we don't silently ignore this and end up with data issues.